tpm2/tpm2.go

```diff
@@ -0,0 +1,1247 @@
+}
+
+// MakeCredential creates an encrypted credential for use in MakeCredential.
+// Returns encrypted credential and wrapped secret used to encrypt it.
+func MakeCredential(rw io.ReadWriter, protectorHandle Handle, credential, activeName []byte) ([]byte, []byte, error) {
+	cmd, err := encodeMakeCredential(protectorHandle, credential, activeName)
+	if err != nil {
+		return nil, nil, err
+	}
+	resp, err := runCommand(rw, tagNoSessions, cmdMakeCredential, tpmutil.RawBytes(cmd))
+	if err != nil {
+		return nil, nil, err
+	}
+	credBlob, encryptedSecret, err := decodeMakeCredential(resp)
+	if err != nil {
+		return nil, nil, fmt.Errorf("decoding MakeCredential response: %v", err)
+	}
+	return credBlob, encryptedSecret, nil
+}
+
+func encodeEvictControl(owner Handle, tmpHandle, persistantHandle Handle) ([]byte, error) {
+	b1, err := tpmutil.Pack(owner)
+	if err != nil {
+		return nil, err
+	}
+	b2, err := tpmutil.Pack(tmpHandle)
+	if err != nil {
+		return nil, err
+	}
+	b3, err := tpmutil.Pack([]byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	b4, err := encodePasswordAuthArea("", HandlePasswordSession)
+	if err != nil {
+		return nil, err
+	}
+	b5, err := tpmutil.Pack(persistantHandle)
+	if err != nil {
+		return nil, err
+	}
+	args := append(b1, b2...)
+	args = append(args, b3...)
+	args = append(args, b4...)
+	args = append(args, b5...)
+	return args, nil
+}
+
+// EvictControl toggles persistence of an object within the TPM.
+func EvictControl(rw io.ReadWriter, owner, tmpHandle, persistantHandle Handle) error {
+	cmd, err := encodeEvictControl(owner, tmpHandle, persistantHandle)
+	if err != nil {
+		return err
+	}
+	_, err = runCommand(rw, tagSessions, cmdEvictControl, tpmutil.RawBytes(cmd))
+	return err
+}
+
+// ContextSave returns an encrypted version of the session, object or sequence
+// context for storage outside of the TPM.
+func ContextSave(rw io.ReadWriter, handle Handle) ([]byte, error) {
+	return runCommand(rw, tagNoSessions, cmdContextSave, handle)
+}
+
+func decodeLoadContext(in []byte) (Handle, error) {
+	var handle Handle
+	err := tpmutil.Unpack(in, &handle)
+	if err != nil {
+		return 0, err
+	}
+	return Handle(handle), nil
+}
+
+// ContextLoad reloads context data created by ContextSave.
+func ContextLoad(rw io.ReadWriter, saveArea []byte) (Handle, error) {
+	resp, err := runCommand(rw, tagNoSessions, cmdContextLoad, tpmutil.RawBytes(saveArea))
+	if err != nil {
+		return 0, err
+	}
+	handle, err := decodeLoadContext(resp)
+	if err != nil {
+		return 0, fmt.Errorf("decoding LoadContext response: %v", err)
+	}
+	return handle, nil
+}
+
+func encodeIncrementNv(handle Handle, authString string) ([]byte, error) {
+	auth, err := encodePasswordAuthArea(authString, HandlePasswordSession)
+	if err != nil {
+		return nil, err
+	}
+	out, err := tpmutil.Pack(handle, handle, []byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	out = append(out, auth...)
+	return out, nil
+}
+
+// NVIncrement increments a counter NV index.
```

Comment:
For clarity, maybe replace "a counter NV index" with "a counter in NVRAM".

Reply:
Done